### PR TITLE
Enhance creation of RateLimitPolicy around missing limits.

### PR DESCRIPTION
**Pull request by Boomatang**
Ensure the user can create a ratelimit policy without at least one limit.

closes: #1023 

# Validate

Try apply the following policies to the installation. They should both fail with different failure messages. Any policy where the name is `toystore-*1` should be applied to the cluster.
```
apiVersion: kuadrant.io/v1
kind: RateLimitPolicy
metadata:
  name: toystore-s1
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: simple-toystore
  limits: 
    "limit-per-ip":
      rates:
      - limit: 5
        window: '10s'
---
apiVersion: kuadrant.io/v1
kind: RateLimitPolicy
metadata:
  name: toystore-s2
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: simple-toystore
---
apiVersion: kuadrant.io/v1
kind: RateLimitPolicy
metadata:
  name: toystore-s3
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: simple-toystore
  limits: {} 
---
apiVersion: kuadrant.io/v1
kind: RateLimitPolicy
metadata:
  name: toystore-d1
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: simple-toystore
  defaults:
    limits: 
      "limit-per-ip":
        rates:
        - limit: 5
          window: '10s'
---
apiVersion: kuadrant.io/v1
kind: RateLimitPolicy
metadata:
  name: toystore-d2
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: simple-toystore
  defaults: {}
---
apiVersion: kuadrant.io/v1
kind: RateLimitPolicy
metadata:
  name: toystore-d3
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: simple-toystore
  defaults:
    limits: {}
---
apiVersion: kuadrant.io/v1
kind: RateLimitPolicy
metadata:
  name: toystore-o1
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: simple-toystore
  overrides:
    limits: 
      "limit-per-ip":
        rates:
        - limit: 5
          window: '10s'
---
apiVersion: kuadrant.io/v1
kind: RateLimitPolicy
metadata:
  name: toystore-o2
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: simple-toystore
  overrides: {}
---
apiVersion: kuadrant.io/v1
kind: RateLimitPolicy
metadata:
  name: toystore-o3
  namespace: default
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: simple-toystore
  overrides:
    limits:  {}
---

```

### CODE REVIEW

File: api/v1/ratelimitpolicy_types.go [hunk 1]
```diff
@@ -153,6 +153,9 @@ func (p *RateLimitPolicy) Kind() string {
 // +kubebuilder:validation:XValidation:rule="!(has(self.defaults) && has(self.limits))",message="Implicit and explicit defaults are mutually exclusive"
 // +kubebuilder:validation:XValidation:rule="!(has(self.defaults) && has(self.overrides))",message="Overrides and explicit defaults are mutually exclusive"
 // +kubebuilder:validation:XValidation:rule="!(has(self.overrides) && has(self.limits))",message="Overrides and implicit defaults are mutually exclusive"
+// +kubebuilder:validation:XValidation:rule="!(has(self.overrides) || has(self.defaults)) ? has(self.limits) && size(self.limits) > 0 : true",message="At least one spec.limits most be defined"
+// +kubebuilder:validation:XValidation:rule="has(self.overrides) ? has(self.overrides.limits) && size(self.overrides.limits) > 0 : true",message="At least one spec.overrides.limits most be defined"
```
Comment (eguzki) [inline api/v1/ratelimitpolicy_types.go:157]:
```suggestion
// +kubebuilder:validation:XValidation:rule="!has(self.overrides) || has(self.overrides.limits) && size(self.overrides.limits) > 0",message="At least one spec.overrides.limits most be defined"
```

Just a matter of personal taste, up to you.

**Review by eguzki — COMMENTED** [review]
I miss a integration test (with actual k8s running) in `tests/common/ratelimitpolicy/ratelimitpolicy_controller_test.go` for the following use cases:
* no defaults, no overrides, no limits -> client return error on create
* with defaults, no limits -> client return error on create
* with overrides, no limits -> client return error on create

File: api/v1/ratelimitpolicy_types.go (hunk 1 above)
Comment (Boomatang) [reply api/v1/ratelimitpolicy_types.go:157]: The ternary operator makes more sense to me. The not has always trips me up. I tend not to like to have navigate checks.

**Comment by Boomatang** [comment]
> I miss a integration test (with actual k8s running) in `tests/common/ratelimitpolicy/ratelimitpolicy_controller_test.go` for the following use cases:

I can add some test no problem.

File: api/v1/ratelimitpolicy_types.go (hunk 1 above)
Comment (eguzki) [reply api/v1/ratelimitpolicy_types.go:157]: LGTM

**Pull request merged**